docs/src/main/asciidoc/class-loading-reference.adoc

```diff
@@ -0,0 +1,163 @@
+////
+= Quarkus - Class Loading Reference
+
+include::./attributes.adoc[]
+
+This document explains the Quarkus class loading architecture. It is intended for extension
+authors and advanced users who want to understand exactly how Quarkus works.
+
+The Quarkus class loading architecture is slightly different depending on the mode that
+the application is run in. When running a production application everything is loaded
+in the system ClassLoader, so it is a completely flat class path. This also applies to
+native image mode which does not really support multiple ClassLoaders, and is based on
+a normal production Quarkus application.
+
+For all other use cases (e.g. tests, dev mode, and building the application) Quarkus
+uses the class loading architecture outlined here.
+
+== Bootstrapping Quarkus
+
+All Quarkus applications are created by the QuarkusBootstrap class in the `independent-projects/bootstrap` module. This
+class is used to resolve all the relevant dependencies (both deployment and runtime) that are needed for the Quarkus
+application. The end result of this process is a `CuratedApplication`, which contains all the class loading information
+for the application.
+
+The `CuratedApplication` can then be used to create an `AugmentAction` instance, which can create production application
+and start/restart runtime ones. This application instance exists within an isolated ClassLoader, it is not necessary
+to have any of the Quarkus deployment classes on the class path as the curate process will resolve them for you.
+
+This bootstrap process should be the same no matter how Quarkus is launched, just with different parameters passed in.
+
+=== Current Run Modes
+
+At the moment we have the following use cases for bootstrapping Quarkus:
+
+- Maven creating production application
+- Maven dev mode
+- Gradle creating a production application
+- Gradle dev mode
+- QuarkusTest (Maven, Gradle and IDE)
+- QuarkusUnitTest (Maven, Gradle and IDE)
+- QuarkusDevModeTest (Maven, Gradle and IDE)
+- Arquillian Adaptor
+
+One of the goals of this refactor is to have all these different run modes boot Quarkus in fundamentally the same way.
+
+== ClassLoader Implementations
+
+Quarkus has the following ClassLoaders:
+
+Base ClassLoader:: 
+
+This is usually the normal JVM System ClassLoader. In some environments such as Maven it may be different. This ClassLoader
+is used to load the bootstrap classes, and other ClassLoader instances will delegate the loading of JDK classes to it.
+
+Augment ClassLoader::
+
+This loads all the `-deployment` artifacts and their dependencies, as well as other user dependencies. It does not load the
+application root or any hot deployed code. This ClassLoader is persistent, even if the application restarts it will remain
+(which is why it cannot load application classes that may be hot deployed). It's parent is the base ClassLoader, and it is
+transformer safe.
```

Comment:
I think we should have an explanation of what that means.

Reply:
The explanation exists, but it's further down, so I agree it should be moved up here

Reply:
Can you also please add an explanation why this classloader needs to be transformer safe? Is this need driven by the fact that we need to transform dependencies in some cases?